### PR TITLE
New block: Post comments link

### PR DESCRIPTION
**Pull request by carolinan**
## Description

Part of https://github.com/WordPress/gutenberg/issues/22724
Adds a new post comments link block.

The block displays the number of comments, and links to the comments area on the single view.

❓  **-It tries to link to #comments even if the single view does not have a comment(s) block.**

## How has this been tested?

I have added the block inside the post content, and viewed the post on the front:
With and without comments, with and without comments enabled.

I have added the block to the index.html file of TT1 Blocks, to make sure that it inherits the context and displays correctly in a query loop.

## Screenshots 
TT1 Blocks, front:
image
_The title is showing because the theme does not have a screen-reader-text class._

## Types of changes

New block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by carolinan** [comment]
I haven't learnt how to write the tests yet, saving that for tomorrow 😸

**Comment by scruffian** [comment]
This looking good. I think it would be good to have an option to customize the text - I might like it to say "Leave a comment" rather than show the number.

It might also be nice to add an option to display an icon next to the link.

**Comment by jasmussen** [comment]
This is what I see:

Screenshot 2021-04-05 at 09 19 18

We need a new icon. But I think that's probably something that can wait, and not something that should block this PR. I'd also echo Ben's comment above about a custom phrasing, but also suggest that's something that can wait (I'd like the same customizability for the pagination blocks, for example).

One thing I noticed is that the block isn't inert:

focusable

I could swear there was precedence in some existing block for making links like that inert using a `Disabled` component, but I can't find it now, and I'm noticing also Latest Posts and others are interactive. So probably that should also not be a blocker.

So I guess this just needs a code review. But would be good to ticket followups. Thank you!

**Comment by carolinan** [comment]
If the disabled component is used we won't be able to also have an editable text.

**Comment by jasmussen** [comment]
> If the disabled component is used we won't be able to also have an editable text.

Good point, though we could just put a textfield in the sidebar. While normally putting it directly in the canvas is the best strategy, I have a vague instinct that the sidebar might be where people intuit a rename to be. But this is not a strong opinion, and as noted, it could definitely be a followup.

**Comment by carolinan** [comment]
The use of links inside the editors may need a more comprehensive review so that the behavior is the same in different blocks.
And if there already are best practices, they need to be documented.
The problem with having the link clickable here is that in the site editor, the linked page is opened inside the iframe.
It can also be confusing with visible links that are not clickable (but can be opened with right click), but it seems like the least bad option.

**Comment by jasmussen** [comment]
Agreed. We've been talking about a "navigation cursor" you can select from the tools drop-down, which would navigate to link destinations inside the editor, when links are clicked. Sometimes referred to as "browse mode".

File: packages/block-library/src/post-comments-link/edit.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+			commentsText = __( 'No comments' );
+		} else if ( commentsCount.length === 1 ) {
+			commentsText = __( 'One comment' );
+		} else {
+			commentsText = sprintf(
+				/* translators: %d Number of comments */
+				__( '%d comments' ),
+				commentsCount
+			);
+		}
```
Comment (aristath) [inline packages/block-library/src/post-comments-link/edit.js]: Shouldn't this be using `_n()`? That would account for languages where the plural is not the same for all numbers (in some languages the plural for 2 is different)

File: packages/block-library/src/post-comments-link/index.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @param  string   $content    Block default content.
+ * @param  WP_Block $block      Block instance.
+ * @return string   Returns the rendered link.
+ */
+function render_block_core_post_comments_link( $attributes, $content, $block ) {
+	if ( ! isset( $block->context['postId'] )
+		|| isset( $block->context['postId'] )
+		&& ! comments_open( $block->context['postId'] ) ) {
+			return '';
+	}
```
Comment (aristath) [inline packages/block-library/src/post-comments-link/index.php]: Nitpicking, but could we change the alignment of this to make things line-up a bit more, and have `return` nested 1 level less?

```suggestion
	if (
		! isset( $block->context['postId'] ) ||
		isset( $block->context['postId'] ) &&
		! comments_open( $block->context['postId'] )
	) {
		return '';
	}
```

File: packages/block-library/src/post-comments-link/index.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+			$post_title
+		);
+	} else {
+		$comment_text = sprintf(
+			/* translators: %1$d Number of comments, %2$s post title */
+			__( '%1$d comments<span class="screen-reader-text"> on %2$s</span>' ),
+			$comments_number,
+			$post_title
+		);
+	}
```
Comment (aristath) [inline packages/block-library/src/post-comments-link/index.php]: This should probably be using `_n()` too.

File: packages/block-library/src/post-comments-link/index.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	return sprintf( '<div %1$s>', $wrapper_attributes ) .
+		'<a href="' . $comments_link . '">' . $comment_text . '</a>' .
+		'</div>';
```
Comment (aristath) [inline packages/block-library/src/post-comments-link/index.php]: A bit of a nitpick, but `sprintf()` is not really necessary here, and we should avoid concatenating strings like `'</a>' . '</div>'` when they can be combined.

```suggestion
	return "<div {$wrapper_attributes}><a href='{$comments_link}'>{$comment_text}</a></div>";
```

**Review by aristath — CHANGES_REQUESTED** [review]

File: packages/block-library/src/post-comments-link/edit.js (hunk 1 above)
Comment (carolinan) [reply packages/block-library/src/post-comments-link/edit.js:80]: Did you mean like this?
```
if ( commentsCount.length === 0 ) {
		commentsText = __( 'No comments' );
	} else {
		commentsText = sprintf(
		/* translators: %d Number of comments */
		_n(
			'%d comment',
			'%d comments',
			commentsCount
		),
		commentsCount
	);
}

```

Comment (carolinan) [reply packages/block-library/src/post-comments-link/edit.js]: Ok that did not work because it now says "1 comments"

Comment (carolinan) [reply packages/block-library/src/post-comments-link/edit.js:80]: The `commentsCount.length i`s also incorrect here

File: packages/block-library/src/post-comments-link/edit.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		if ( commentsNumber === 0 ) {
+			commentsText = __( 'No comments' );
+		} else {
+			commentsText = sprintf(
+				/* translators: %s: Number of comments */
+				_n(
+					'%s comment',
+					'%s comments',
+					commentsNumber
+				),
```
Comment (aristath) [inline packages/block-library/src/post-comments-link/edit.js]: The linter is complaining, this needs to change to the below:
```suggestion
				_n( '%s comment', '%s comments', commentsNumber ),
```

**Review by aristath — APPROVED** [review]

**Pull request merged**